File: go-backend-interview-prep/06_concurrency/08_mutexes/main.go
package main

import (
	"fmt"
	"sync"
)

// --- Example 1: safe counter with Mutex ---

type SafeCounter struct {
	mu    sync.Mutex
	value int
}

func (c *SafeCounter) Inc() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value++
}

func (c *SafeCounter) Value() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.value
}

// --- Example 2: read-heavy cache with RWMutex ---

type Cache struct {
	mu   sync.RWMutex
	data map[string]string
}

func NewCache() *Cache {
	return &Cache{data: make(map[string]string)}
}

func (c *Cache) Get(key string) (string, bool) {
	c.mu.RLock() // shared read lock
	defer c.mu.RUnlock()
	val, ok := c.data[key]
	return val, ok
}

func (c *Cache) Set(key, val string) {
	c.mu.Lock() // exclusive write lock
	defer c.mu.Unlock()
	c.data[key] = val
}

func main() {
	// --- Safe counter ---
	fmt.Println("=== Safe Counter (Mutex) ===")
	counter := &SafeCounter{}
	var wg sync.WaitGroup

	for i := 0; i < 1000; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			counter.Inc()
		}()
	}
	wg.Wait()
	fmt.Println("  counter:", counter.Value()) // always 1000

	// --- RWMutex cache ---
	fmt.Println("\n=== Cache (RWMutex) ===")
	cache := NewCache()
	cache.Set("name", "alice")
	cache.Set("role", "engineer")

	// Multiple readers can run concurrently
	var wg2 sync.WaitGroup
	for i := 0; i < 5; i++ {
		wg2.Add(1)
		go func(id int) {
			defer wg2.Done()
			if val, ok := cache.Get("name"); ok {
				fmt.Printf("  reader %d: name=%s\n", id, val)
			}
		}(i)
	}
	wg2.Wait()

	// Write blocks all readers
	cache.Set("name", "bob")
	val, _ := cache.Get("name")
	fmt.Println("  after write: name =", val)
}

File: go-backend-interview-prep/06_concurrency/08_mutexes/main.py
"""Mutexes -- Python equivalent of the Go example.

Python's GIL protects some built-in operations, but compound
operations (read-modify-write) still need threading.Lock.
"""

import threading


# --- Example 1: safe counter with Lock ---

class SafeCounter:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._value = 0

    def inc(self) -> None:
        with self._lock:  # auto lock/unlock (like defer Unlock)
            self._value += 1

    def value(self) -> int:
        with self._lock:
            return self._value


# --- Example 2: read-heavy cache with RLock ---
# Python has no RWLock in stdlib. threading.RLock is reentrant,
# not a read-write lock. We use a regular Lock here.

class Cache:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._data: dict[str, str] = {}

    def get(self, key: str) -> str | None:
        with self._lock:
            return self._data.get(key)

    def set(self, key: str, val: str) -> None:
        with self._lock:
            self._data[key] = val


def main() -> None:
    # --- Safe counter ---
    print("=== Safe Counter (Lock) ===")
    counter = SafeCounter()
    threads = []

    for _ in range(1000):
        t = threading.Thread(target=counter.inc)
        threads.append(t)
        t.start()

    for t in threads:
        t.join()
    print(f"  counter: {counter.value()}")  # always 1000

    # --- Cache ---
    print("\n=== Cache (Lock) ===")
    cache = Cache()
    cache.set("name", "alice")
    cache.set("role", "engineer")

    threads = []
    for i in range(5):
        def reader(rid):
            val = cache.get("name")
            if val:
                print(f"  reader {rid}: name={val}")

        t = threading.Thread(target=reader, args=(i,))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    cache.set("name", "bob")
    print("  after write: name =", cache.get("name"))


if __name__ == "__main__":
    main()
